flight/Libraries/math/cic.c

```diff
@@ -0,0 +1,160 @@
+			filter_data->combState0_old = filter_data->combState0;
+		}
+		else if (cic_filter_ddelay == 2) {
+			filter_data->integrateState_old      = filter_data->integrateState_old_temp;
+			filter_data->integrateState_old_temp = filter_data->integrateState1;
+			filter_data->combState0_old      = filter_data->combState0_old_temp;
+			filter_data->combState0_old_temp = filter_data->combState0;
+		}
+	}
+
+	else if (cic_filter_order == 3) {
+		filter_data->combState0 = filter_data->integrateState2 - filter_data->integrateState_old;
+		filter_data->combState1 = filter_data->combState0  - filter_data->combState0_old;
+		filter_data->combState2 = filter_data->combState1  - filter_data->combState1_old;
+
+		if (cic_filter_ddelay == 1) {
+			filter_data->integrateState_old = filter_data->integrateState2;
+			filter_data->combState0_old = filter_data->combState0;
+			filter_data->combState1_old = filter_data->combState1;
+		}
+		else if (cic_filter_ddelay == 2) {
+			filter_data->integrateState_old      = filter_data->integrateState_old_temp;
+			filter_data->integrateState_old_temp = filter_data->integrateState2;
+			filter_data->combState0_old      = filter_data->combState0_old_temp;
+			filter_data->combState0_old_temp = filter_data->combState0;
+			filter_data->combState1_old      = filter_data->combState1_old_temp;
+			filter_data->combState1_old_temp = filter_data->combState1;
+		}
+	}
+}
+
+/**
+ * @brief Get the decimated / down-sampled output of the CIC filter
+ * @param[in] filter_data Pointer to the CIC filter structure which stores the data / information
+ * @returns Output the computed gain of the filter
+ */
+int32_t cic_get_decimation_output(struct cic_filter_data *filter_data) {
+	if (cic_filter_order == 1)
+		return filter_data->combState0;
+
+	else if (cic_filter_order == 2)
+		return filter_data->combState1;
+
+	else if (cic_filter_order == 3)
+		return filter_data->combState2;
+
+	else // (cic_filter_order == 0)
+		return filter_data->integrateState0;
+}
+
+
+/**
+ * @brief Get the gain of the CIC filter, e.g. for normalisation
+ * @returns Output the computed gain of the filter
+ */
+float cic_get_gain(void) {
+	// CIC:
+	// Differential Delay N = D/R = Delay / Downsampling factor
+	// Gain: (N*R)^M
+	if (cic_filter_order == 0) // simple boxcar averaging
+		return (cic_filter_decimation);
+	else
+		return (pow((cic_filter_decimation * cic_filter_ddelay), cic_filter_order));
+}
+
+
+/**
+ * @brief Configure the common parameters of the CIC filter
+ * @param[in] cic_filter_order The order of the CIC filter
+ * @param[in] cic_filter_ddelay  The differential delay of the CIC filter
+ * @param[in] cic_filter_ddelay  The decimation factor of the CIC filter
+ */
+void cic_configure(uint8_t filter_order, uint8_t filter_ddelay,  uint8_t filter_decimation) {
+	cic_filter_order     = bound_min_max(filter_order, 0, 3);
+	cic_filter_ddelay    = bound_min_max(filter_ddelay, 1, 2);
+	cic_filter_decimation= bound_min_max(filter_decimation, 1, 255);
```

Comment:
> But what should we do if a dynamic check failed?

Right now you're bounding it already within what we consider "sane" parameters.  We can turn down a user's parameter to be within the gain limit.

Ideally we'd also assert an alarm, but the infrastructure for that is currently not so good (to communicate configuration errors).